File: models/Game.ts
import mongoose, { Schema, Document } from 'mongoose'
import type { GameSettings, SongEntry } from '@/types'

const SongEntrySchema = new Schema<SongEntry>({
  annId: Number,
  songName: { type: String, required: true },
  songArtist: { type: String, required: true },
  animeName: { type: String, required: true },
  animeVintage: String,
  songType: String,
  videoUrl: { type: String, required: true },
}, { _id: false })

const CardSchema = new Schema({
  label: { type: String, required: true },
  stars: { type: Number, min: 1, max: 5, required: true },
  songs: [SongEntrySchema],
})

const GameSettingsSchema = new Schema<GameSettings>({
  clipDuration: { type: Number, default: 60 },
  wrongAnswerDeduction: { type: Number, default: 0 },
  starPointMap: {
    type: Map,
    of: Number,
    default: { 1: 100, 2: 200, 3: 300, 4: 400, 5: 500 },
  },
}, { _id: false })

export interface IGame extends Document {
  name: string
  settings: GameSettings
  cards: mongoose.Types.DocumentArray<mongoose.Document & { label: string; stars: number; songs: SongEntry[] }>
  createdAt: Date
  updatedAt: Date
}

const GameSchema = new Schema<IGame>(
  {
    name: { type: String, required: true },
    settings: { type: GameSettingsSchema, default: () => ({}) },
    cards: [CardSchema],
  },
  { timestamps: true }
)

export default mongoose.models.Game || mongoose.model<IGame>('Game', GameSchema)

File: backend/Models/Game.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Bson.Serialization.Options;

namespace SpeedTune.Api.Models;

public class SongEntry
{
    public int AnnId { get; set; }
    public string SongName { get; set; } = "";
    public string SongArtist { get; set; } = "";
    public string AnimeName { get; set; } = "";
    public string AnimeVintage { get; set; } = "";
    public string SongType { get; set; } = "";
    public string VideoUrl { get; set; } = "";
    /// <summary>Song duration in seconds (from AnisongDB), used to calculate playback range.</summary>
    public float SongLength { get; set; } = 0;
    /// <summary>Minimum playback start point as a percentage 0–100.</summary>
    public int StartMin { get; set; } = 0;
    /// <summary>Maximum playback start point as a percentage 0–100.</summary>
    public int StartMax { get; set; } = 0;
}

public class Card
{
    // Cards are embedded documents — store Id as a plain string (UUID from frontend is fine)
    public string Id { get; set; } = ObjectId.GenerateNewId().ToString();
    public string Label { get; set; } = "";
    public int Stars { get; set; } = 1;
    public List<SongEntry> Songs { get; set; } = new();
}

public class GameSettings
{
    public int ClipDuration { get; set; } = 60;
    public int BuzzTimerSeconds { get; set; } = 30;
    public int WrongAnswerDeduction { get; set; } = 0;
    [BsonDictionaryOptions(DictionaryRepresentation.ArrayOfArrays)]
    public Dictionary<int, int> StarPointMap { get; set; } = new()
    {
        { 1, 100 }, { 2, 200 }, { 3, 300 }, { 4, 400 }, { 5, 500 }
    };
}

public class Game
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = ObjectId.GenerateNewId().ToString();
    public string Name { get; set; } = "";
    public GameSettings Settings { get; set; } = new();
    public List<Card> Cards { get; set; } = new();
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}
